src/main/java/org/broadinstitute/consent/http/service/DatasetRegistrationService.java

```diff
@@ -0,0 +1,426 @@
+        }
+
+        return consentGroupFSOs;
+
+    }
+
+    private FileStorageObject uploadFile(Map<String, FormDataBodyPart> files,
+                                         Map<String, BlobId> uploadedFileCache,
+                                         User user,
+                                         String name,
+                                         FileCategory category) throws IOException {
+
+        FormDataBodyPart bodyPart = files.get(name);
+
+        if (!uploadedFileCache.containsKey(name)) {
+            BlobId id = uploadFile(bodyPart);
+            uploadedFileCache.put(name, id);
+        }
+
+        BlobId id = uploadedFileCache.get(name);
+
+        FileStorageObject fso = new FileStorageObject();
+        fso.setCategory(category);
+        fso.setFileName(bodyPart.getContentDisposition().getFileName());
+        fso.setMediaType(bodyPart.getMediaType().toString());
+        fso.setBlobId(id);
+        fso.setCreateUserId(user.getUserId());
+
+        return fso;
+    }
+
+
+    /**
+     * Extracts an individual field as a dataset property.
+     *
+     * @param name The human-readable name of the field
+     * @param schemaProp The schema property name (camelCase)
+     * @param type The type of the field, e.g. Boolean, String
+     * @param getField Lambda which gets the field's value
+     */
+    private record DatasetPropertyExtractor(
+            String name,
+            String schemaProp,
+            DatasetPropertyType type,
+
+            /*
+             * Takes in: Dataset registration object and consent group index (as integer)
+             * Produces: The value of the field, can be null if field not present.
+             */
+            BiFunction<DatasetRegistrationSchemaV1,Integer,Object> getField
+    ) {
+
+        /**
+         * Converts a field on the given registration to a DatasetProperty.
+         *
+         * @param registration The registration object to extract from
+         * @param consentGroupIdx The index of the consent group to extract from
+         * @return The dataset property, if the field has a value, otherwise Optional.empty()
+         */
+        Optional<DatasetProperty> extract(DatasetRegistrationSchemaV1 registration, int consentGroupIdx) {
+            Object value = this.getField.apply(registration, consentGroupIdx);
+            if (Objects.isNull(value)) {
+                return Optional.empty();
+            }
+
+            DatasetProperty datasetProperty = new DatasetProperty();
+            datasetProperty.setPropertyName(this.name);
+            datasetProperty.setPropertyType(this.type);
+            datasetProperty.setSchemaProperty(this.schemaProp);
+            datasetProperty.setPropertyValue(this.type.coerce(value.toString()));
+
+            return Optional.of(datasetProperty);
+
+        }
+    };
+
+    private static final List<DatasetPropertyExtractor> datasetPropertyExtractors = List.of(
```

Comment:
There's a common pattern to have static finals be in snake case.  Not sure if this a pattern for you (e.g. 
```suggestion
    private static final List<DatasetPropertyExtractor> DATASET_PROPERTY_EXTRACTORS = List.of(
```